openedx/core/djangoapps/programs/utils.py

```diff
@@ -105,33 +103,38 @@ def get_programs_for_credentials(user, programs_credentials):
     return certificate_programs
 
 
-def get_user_enrolled_programs(user, course_keys):
-    '''
-    The helper function that takes in user enrollment information,
-    and returns the list of programs the user is enrolled in.
+def get_engaged_programs(user, enrollments):
+    """Derive a list of programs in which the given user is engaged.
+
     Arguments:
-        user (User): The user to authenticate as for requesting programs.
-        course_key (list): List of courses keys of courses user is enrolled in
+        user (User): The user for which to find programs.
+        enrollments (list): The user's enrollments.
 
     Returns:
-        list, contains programs enrolled
-    '''
-    # Get the programs by courses dictionary from the get_programs_for_dashboard function call
-    programs_dict = get_programs_for_dashboard(
-        user,
-        course_keys)
-    program_list = []
-    # Extract all the programs from the dictionary above
-    for programs_by_courses in programs_dict.values():
-        for program in programs_by_courses:
-            #Remove duplicate programs through the if check below
-            if program not in program_list:
-                program_list.append(program)
-
-    for program in program_list:
-        program['marketing_url'] = urljoin(
-            settings.MKTG_URLS.get('ROOT'),
-            'xseries' + '/{}'
-        ).format(program['marketing_slug'])
-
-    return program_list
+        list of serialized programs, ordered by most recent enrollment
+    """
+    programs = get_programs(user)
+
+    enrollments = sorted(enrollments, key=lambda e: e.created, reverse=True)
+    # enrollment.course_id is really a course key.
+    course_ids = [unicode(e.course_id) for e in enrollments]
+
+    flattened = []
+    for program in programs:
+        run_ids = [
+            run['course_key']
+            for course_code in program['course_codes'] for run in course_code['run_modes']
+        ]
+
+        flattened.append(
+            (program, run_ids)
+        )
+
+    engaged_programs = []
+    for course_id in course_ids:
+        for flat in flattened:
+            program, run_ids = flat
```

Comment:
optional: instead of nested loops, consider inverting the `flattened` structure, so it contains lists of programs keyed by run_id, then join those lists in the order determined by `course_ids`